core/src/syscall/precompiles/sha512/compress/air.rs

```diff
@@ -0,0 +1,484 @@
+        let d = helper(local, 3);
+        let e = helper(local, 4);
+        let f = helper(local, 5);
+        let g = helper(local, 6);
+        let h = helper(local, 7);
+
+        // S1 := (e rightrotate 14) xor (e rightrotate 18) xor (e rightrotate 41)
+        // Calculate e rightrotate 14.
+        FixedRotateRight64Operation::<AB::F>::eval(
+            builder,
+            e,
+            14,
+            local.e_rr_14,
+            local.shard,
+            &local.channel,
+            local.is_real,
+        );
+        // Calculate e rightrotate 18.
+        FixedRotateRight64Operation::<AB::F>::eval(
+            builder,
+            e,
+            18,
+            local.e_rr_18,
+            local.shard,
+            &local.channel,
+            local.is_real,
+        );
+        // Calculate e rightrotate 41.
+        FixedRotateRight64Operation::<AB::F>::eval(
+            builder,
+            e,
+            41,
+            local.e_rr_41,
+            local.shard,
+            &local.channel,
+            local.is_real,
+        );
+        // Calculate (e rightrotate 14) xor (e rightrotate 18).
+        Xor64Operation::<AB::F>::eval(
+            builder,
+            local.e_rr_14.value,
+            local.e_rr_18.value,
+            local.s1_intermediate,
+            local.shard,
+            &local.channel,
+            local.is_real,
+        );
+        // Calculate S1 := ((e rightrotate 14) xor (e rightrotate 18)) xor (e rightrotate 41).
+        Xor64Operation::<AB::F>::eval(
+            builder,
+            local.s1_intermediate.value,
+            local.e_rr_41.value,
+            local.s1,
+            local.shard,
+            &local.channel,
+            local.is_real,
+        );
+
+        // Calculate ch := (e and f) xor ((not e) and g).
+        // Calculate e and f.
+        And64Operation::<AB::F>::eval(
+            builder,
+            e,
+            f,
+            local.e_and_f,
+            local.shard,
+            local.channel,
+            local.is_real,
+        );
+        // Calculate not e.
+        Not64Operation::<AB::F>::eval(
+            builder,
+            e,
+            local.e_not,
+            local.shard,
+            local.channel,
+            local.is_real,
+        );
+        // Calculate (not e) and g.
+        And64Operation::<AB::F>::eval(
+            builder,
+            local.e_not.value,
+            g,
+            local.e_not_and_g,
+            local.shard,
+            local.channel,
+            local.is_real,
+        );
+        // Calculate ch := (e and f) xor ((not e) and g).
+        Xor64Operation::<AB::F>::eval(
+            builder,
+            local.e_and_f.value,
+            local.e_not_and_g.value,
+            local.ch,
+            local.shard,
+            &local.channel,
+            local.is_real,
+        );
+
+        // Calculate temp1 := h + S1 + ch + k[i] + w[i].
+        Add64Operation::<AB::F>::eval(
```

Comment:
Just making a note, but the `Add` operation does not need to range check the inputs since we can assume they are bytes (either as an XOR result or because they come from the ZKVM)